File: backend/tracker-sdk/src/main/java/com/gateflow/sdk/auth/AuthClient.java
package com.gateflow.sdk.auth;

import com.fasterxml.jackson.databind.ObjectMapper;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Exchanges an appKey for a JWT token from the tracker-service auth endpoint.
 * Token is cached and auto-refreshed before expiry.
 */
public class AuthClient {

    private final String authUrl;
    private final String appKey;
    private final HttpClient http;
    private final ObjectMapper mapper;

    private volatile String token;
    private volatile long expiresAt; // epoch millis

    public AuthClient(String baseUrl, String appKey) {
        this.authUrl = baseUrl.replaceAll("/$", "") + "/api/v1/collect/auth";
        this.appKey = appKey;
        this.http = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(5))
                .build();
        this.mapper = new ObjectMapper();
    }

    /**
     * Returns a valid token, refreshing if absent or within 60s of expiry.
     */
    public String getToken() {
        if (token != null && System.currentTimeMillis() < expiresAt - 60_000) {
            return token;
        }
        synchronized (this) {
            if (token != null && System.currentTimeMillis() < expiresAt - 60_000) {
                return token;
            }
            try {
                HttpRequest req = HttpRequest.newBuilder()
                        .uri(URI.create(authUrl))
                        .header("Content-Type", "application/json")
                        .header("X-App-Key", appKey)
                        .POST(HttpRequest.BodyPublishers.ofString("{}"))
                        .timeout(Duration.ofSeconds(10))
                        .build();

                HttpResponse<String> resp = http.send(req, HttpResponse.BodyHandlers.ofString());
                if (resp.statusCode() == 200) {
                    @SuppressWarnings("unchecked")
                    Map<String, Object> body = mapper.readValue(resp.body(), Map.class);
                    token = (String) body.get("sdkToken");
                    Object exp = body.get("expiresIn");
                    long expiresIn = exp instanceof Number ? ((Number) exp).longValue() : 3600;
                    expiresAt = System.currentTimeMillis() + expiresIn * 1000;
                    return token;
                }
                throw new RuntimeException("Auth failed: HTTP " + resp.statusCode() + " " + resp.body());
            } catch (RuntimeException e) {
                throw e;
            } catch (Exception e) {
                throw new RuntimeException("Auth failed: " + e.getMessage(), e);
            }
        }
    }

    /** Async version. */
    public CompletableFuture<String> getTokenAsync() {
        return CompletableFuture.supplyAsync(this::getToken);
    }

    /** Force token refresh on next getToken(). */
    public void invalidate() {
        synchronized (this) {
            token = null;
            expiresAt = 0;
        }
    }
}

File: tracker-sdk/src/auth/AuthClient.ts
/**
 * Exchanges an appKey for a JWT token from the tracker-service auth endpoint.
 * Token is cached and auto-refreshed before expiry.
 *
 * Mirrors the Java SDK {@code com.gateflow.sdk.auth.AuthClient} behavior:
 * - POST {baseUrl}/api/v1/collect/auth with X-App-Key header
 * - Caches {sdkToken, expiresIn} response
 * - Refreshes 60s before expiry
 * - Invalidates on 401 response
 */
export interface AuthResponse {
  sdkToken: string;
  expiresIn: number;
}

export class AuthClient {
  private readonly authUrl: string;
  private readonly appKey: string;
  private token: string | null = null;
  private expiresAt: number = 0; // epoch millis

  /**
   * @param baseUrl  Base URL of the tracker server
   * @param appKey   Application key for authentication
   */
  constructor(baseUrl: string, appKey: string) {
    this.authUrl = baseUrl.replace(/\/+$/, '') + '/api/v1/collect/auth';
    this.appKey = appKey;
  }

  /**
   * Returns a valid token, refreshing if absent or within 60s of expiry.
   * Throws if the auth request fails.
   */
  async getToken(): Promise<string> {
    if (this.token && Date.now() < this.expiresAt - 60_000) {
      return this.token;
    }

    const response = await fetch(this.authUrl, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-App-Key': this.appKey,
      },
      body: '{}',
    });

    if (!response.ok) {
      throw new Error(`Auth failed: HTTP ${response.status} ${await response.text()}`);
    }

    const body: AuthResponse = await response.json();
    this.token = body.sdkToken;
    this.expiresAt = Date.now() + body.expiresIn * 1000;

    return this.token!;
  }

  /**
   * Synchronous getter — returns the cached token or null.
   * Does NOT trigger a refresh. Use getToken() for guaranteed-fresh token.
   */
  getCachedToken(): string | null {
    if (this.token && Date.now() < this.expiresAt - 60_000) {
      return this.token;
    }
    return null;
  }

  /** Force token refresh on next getToken(). */
  invalidate(): void {
    this.token = null;
    this.expiresAt = 0;
  }
}
